From bdd1e621b7637ef966488f528c50d31dbf89ea3f Mon Sep 17 00:00:00 2001
From: Anju Kunjumon <akunjumon@sierrawireless.com>
Date: Tue, 14 Jan 2020 11:03:47 -0800
Subject: [PATCH] Retry device_open without direct-io

For an unknown reason, it seems that the test for direct-io
in device_ready can pass while the same call in device_open fails.

Retry the device_open without direct-io in that case.

Signed-off-by: Bertrand Roussel <broussel@sierrawireless.com>
Modified-by: Anju Kunjumon <akunjumon@sierrawireless.com>
---
 lib/utils_device.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/lib/utils_device.c b/lib/utils_device.c
index 967b7a8..6bf6b77 100644
--- a/lib/utils_device.c
+++ b/lib/utils_device.c
@@ -280,9 +280,23 @@ static int device_open_internal(struct crypt_device *cd, struct device *device,
 	}
 
 	if (device_locked(device->lh))
+	{
 		devfd = _open_locked(cd, device, flags);
+		if ((devfd == EINVAL) && device->o_direct)
+		{
+			log_dbg("Cannot open device %s with direct-io, retrying without.", device_path(device));
+			devfd = _open_locked(cd, device, flags ^ O_DIRECT);
+		}
+	}
 	else
+	{
 		devfd = open(device_path(device), flags);
+		if ((devfd < 0) && device->o_direct && (errno == EINVAL))
+		{
+			log_dbg("Cannot open device %s with direct-io, retrying without.", device_path(device));
+			devfd = open(device_path(device), flags ^ O_DIRECT);
+		}
+	}
 
 	if (devfd < 0) {
 		log_dbg(cd, "Cannot open device %s%s.",
-- 
2.17.1

